torch/nn/_functions/rnn.py

```diff
@@ -251,95 +251,48 @@ def forward(input, weight, hidden):
-        self.weight_buf = flat_weight
-        if flat_weight is None:
-            warnings.warn("RNN module weights are not part of single contiguous "
-                          "chunk of memory. This means they need to be compacted "
-                          "at every call, possibly greatly increasing memory usage. "
-                          "To compact weights again call flatten_parameters().", stacklevel=5)
-
-    def forward_extended(self, input, weight, hx):
-        assert cudnn.is_acceptable(input)
-        # TODO: raise a warning if weight_data_ptr is None
-
-        output = input.new()
-
-        if torch.is_tensor(hx):
-            hy = hx.new()
+def CudnnRNN(mode, input_size, hidden_size, num_layers=1,
+             batch_first=False, dropout=0, train=True, bidirectional=False,
+             batch_sizes=None, dropout_state=None, flat_weight=None):
+    if dropout_state is None:
+        dropout_state = {}
+    mode = cudnn.rnn.get_cudnn_mode(mode)
+    dropout_seed = torch.IntTensor(1).random_()[0]
+    if flat_weight is None:
+        warnings.warn("RNN module weights are not part of single contiguous "
+                      "chunk of memory. This means they need to be compacted "
+                      "at every call, possibly greatly increasing memory usage. "
+                      "To compact weights again call flatten_parameters().", stacklevel=5)
+
+    def forward(input, weight, hx):
+        if mode == cudnn.CUDNN_LSTM:
+            hx, cx = hx
         else:
-            hy = tuple(h.new() for h in hx)
-
-        cudnn.rnn.forward(self, input, hx, weight, output, hy)
-
-        self.save_for_backward(input, hx, weight, output)
-        return output, hy
-
-    def backward_extended(self, grad_output, grad_hy):
-        input, hx, weight, output = self.saved_tensors
-        input = input.contiguous()
-
-        grad_input, grad_weight, grad_hx = None, None, None
-
-        assert cudnn.is_acceptable(input)
-
-        grad_input = input.new()
-        if torch.is_tensor(hx):
-            grad_hx = input.new()
-        else:
-            grad_hx = tuple(h.new() for h in hx)
-
-        if self.retain_variables:
-            self._reserve_clone = self.reserve.clone()
-
-        cudnn.rnn.backward_grad(
-            self,
-            input,
-            hx,
-            weight,
-            output,
-            grad_output,
-            grad_hy,
-            grad_input,
-            grad_hx)
-
-        if any(self.needs_input_grad[1:]):
-            grad_weight = [tuple(w.new().resize_as_(w) for w in layer_weight) for layer_weight in weight]
-            cudnn.rnn.backward_weight(
-                self,
-                input,
-                hx,
-                output,
-                weight,
-                grad_weight)
+            cx = None
+
+        handle = cudnn.get_handle()
+        dropout_desc = cudnn.rnn.init_dropout_descriptor(handle, dropout, train, dropout_seed, dropout_state)
```

Comment:
Why do we initialize the descriptor here (and at every forward!)? I thought you had a bunch of code dedicated to handling dropout descriptors in C++

Reply:
That's the only place where dropout descriptor is properly initialized, all calls in C++ are to cheap alternatives where you either don't need it, or already have a properly initialized state that just needs to be restored.

Reply:
Actually, I should have ported this to C++ too, so there's that.